Cython/Utility/ModuleSetupCode.c

```diff
@@ -2190,17 +2221,25 @@ static PyObject* __Pyx_PyCode_New(
     }
 
     #if CYTHON_COMPILING_IN_LIMITED_API
+    #if CYTHON_AVOID_THREAD_UNSAFE_BORROWED_REFS && __PYX_LIMITED_VERSION_HEX >= 0x030d00b1
+    if (unlikely(PyDict_GetItemRef(tuple_dedup_map, varnames_tuple, &varnames_tuple_dedup) < 0)) goto done;
+    #else
     varnames_tuple_dedup = PyDict_GetItem(tuple_dedup_map, varnames_tuple);
+    #endif
     if (!varnames_tuple_dedup) {
         if (unlikely(PyDict_SetItem(tuple_dedup_map, varnames_tuple, varnames_tuple) < 0)) goto done;
         varnames_tuple_dedup = varnames_tuple;
     }
     #else
+    #if CYTHON_AVOID_THREAD_UNSAFE_BORROWED_REFS && PY_VERSION_HEX >= 0x030d00b1
+    PyDict_SetDefaultRef(tuple_dedup_map, varnames_tuple, varnames_tuple, &varnames_tuple_dedup);
+    #else
     varnames_tuple_dedup = PyDict_SetDefault(tuple_dedup_map, varnames_tuple, varnames_tuple);
+    #endif
     if (unlikely(!varnames_tuple_dedup)) goto done;
     #endif
 
-    #if CYTHON_AVOID_BORROWED_REFS
+    #if CYTHON_AVOID_BORROWED_REFS && !CYTHON_AVOID_THREAD_UNSAFE_BORROWED_REFS
     Py_INCREF(varnames_tuple_dedup);
```

Comment:
This is not correct since we've applied a version check above. In older Pythons, the reference will not be increfed.

I'd move the error check + incref into the `#else` block above (and duplicate the error check in the `#if` block).